Fix a few dwarf64 parsing problems.  Nothing fancy.

quenelle opened the pull request:
This fixes bugs that show up when dwarf from Studio compilers
gets mixed in via crt files.

Commit:
Fix a few dwarf64 parsing problems.  Nothing fancy.
This fixes bugs that show up when dwarf from Studio compilers
gets mixed in via crt files.

quenelle commented:
More specifically: 
Fixes #25

binarycrusader commented on src/debug/dwarf/open.go:
```diff
@@ -58,7 +58,14 @@ func New(abbrev, aranges, frame, info, line, pubnames, ranges, str []byte) (*Dat
 	x, y := d.info[4], d.info[5]
+        if (d.info[0] == 255 && d.info[1] == 255) {
+           // 64-bit DWARF has 4 bytes of FF, then 8 byte size,
+	   // then 2 byte version
```
The formatting seems a bit off here; consider running 'gofmt -w' on this file.

binarycrusader commented on src/debug/dwarf/unit.go:
```diff
@@ -62,12 +62,23 @@ func (d *Data) parseUnits() ([]unit, error) {
+			aoff = b.uint32()
+			header_bytes_read += 4
+		} else {
+			// For now ignore the high bits
```
s/now/now,/

binarycrusader commented on src/debug/dwarf/unit.go:
```diff
@@ -62,12 +62,23 @@ func (d *Data) parseUnits() ([]unit, error) {
 		n, u.is64 = b.unitLength()
 		vers := b.uint16()
+		var header_bytes_read uint
+		header_bytes_read = 2 // Not counting size
```
Maybe 'hbytes' instead of 'header_bytes_read' or 'hlen' ?

Commit:
Fix a few dwarf64 parsing problems.
This fixes bugs that show up when dwarf from Studio compilers
gets mixed in via crt files.

binarycrusader reviewed (commented):
Do all of the Go tests pass still after this change (e.g. ./all.bash)?

aclements commented on src/debug/dwarf/unit.go:
```diff
@@ -62,12 +62,24 @@ func (d *Data) parseUnits() ([]unit, error) {
+			hbcount += 4
+		} else {
+			// For now, ignore the high bits
+			aoff = uint32(b.uint64())
```
I'd recommend checking for overflow here. See the equivalent code in Data.parseTypes.

aclements commented on src/debug/dwarf/unit.go:
```diff
@@ -62,12 +62,24 @@ func (d *Data) parseUnits() ([]unit, error) {
 		var n Offset
 		n, u.is64 = b.unitLength()
 		vers := b.uint16()
+		// Count the bytes of header read, not counting size
```
Rather than counting bytes, I'd suggest following what parseTypes does. Here, just say "hdroff := b.off", then below say "b.bytes(int(n - (b.off - hdroff)))".

binarycrusader replied:
I had actually suggested the same thing to Chris offline :-)

aclements reviewed (commented):
Looks correct to me. Just a few non-semantic changes.

This would be great to upstream separate of the rest of the sparc changes. I'm really surprised this hasn't bitten us before.

4ad commented:
> This would be great to upstream separate of the rest of the sparc changes. I'm really surprised this hasn't bitten us before.

I agree.

quenelle commented:
Sorry for the delay here.
I'm on medical leave after breaking my hip and wrist at the same time.
It'll be some time before I'm reasonably functional at a keyboard again.
If someone else is interested in finishing this, I would not object.

Changes needed are:

1) Instead of counting header bytes as they are read in, just read b.off 
   to see how many more to skip.  As it is in typeunit.go/parseTypes()

2) Check for overflow of the abbrev table offset in unit.go/New as it's
   done in typeunit.go/parseTypes

3) add a test case. First try using a simple hello world binary from 
   Studio cc in -m64 mode.  If that does fail, then grab the
   frankenstein binary from the current failing test on machines
   with Studio dwarf in the crt files. You can check a sample elf
   binary (minimal) into src/debug/dwarf/testdata/one 

[dw64studiocrt.zip](https://github.com/4ad/go/files/1195675/dw64studiocrt.zip)

I've attached a binary that should show the error if you use DWARF() to open it.

quenelle commented:
I agree pushing this separate from the sparc64 stuff seems fine.
